src/v/cluster/plugin_frontend.h

```diff
@@ -0,0 +1,164 @@
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "cluster/commands.h"
+#include "cluster/partition_leaders_table.h"
+#include "cluster/plugin_table.h"
+#include "cluster/topic_table.h"
+#include "cluster/types.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "model/timeout_clock.h"
+#include "rpc/connection_cache.h"
+#include "seastarx.h"
+#include "utils/mutex.h"
+
+#include <absl/base/thread_annotations.h>
+
+#include <variant>
+
+#pragma once
+
+namespace cluster {
+
+/**
+ * The plugin frontend takes care of all (WASM) plugin related functionality and
+ * servicing requests from the rest of the cluster.
+ *
+ * Instance per core.
+ */
+class plugin_frontend : public ss::peering_sharded_service<plugin_frontend> {
+private:
+    using transform_cmd
+      = std::variant<transform_update_cmd, transform_remove_cmd>;
+
+public:
+    // The plugin frontend takes the corresponding services for it's core.
+    //
+    // The exception being the controller_stm which is only nonnull for the
+    // service on core 0.
+    plugin_frontend(
+      model::node_id,
+      partition_leaders_table*,
+      plugin_table*,
+      topic_table*,
+      controller_stm*,
+      rpc::connection_cache*,
+      ss::abort_source*);
+
+    using notification_id = plugin_table::notification_id;
+    using notification_callback = plugin_table::notification_callback;
+    struct mutation_result {
+        // The UUID of the modified transform, mostly useful for cleanup
+        // of wasm_binaries.
+        // See transform_metadata::uuid
+        uuid_t uuid;
+        errc ec;
+    };
+
+    // Create or update a transform by name.
+    ss::future<errc> upsert_transform(
+      model::transform_metadata, model::timeout_clock::time_point);
+
+    // Remove a transform by name.
+    ss::future<mutation_result>
+      remove_transform(model::transform_name, model::timeout_clock::time_point);
+
+    // Register for updates going forward.
+    //
+    // This is a core local operation, you can register for updates to the
+    // transforms table. If you want to lookup the transform you can then do
+    // that in the callback.
+    notification_id register_for_updates(notification_callback);
+
+    // Unregister a previously registered notification.
+    void unregister_for_updates(notification_id);
+
+    // Lookup a transform by ID. Useful when paired with registering for
+    // updates.
+    std::optional<model::transform_metadata>
+      lookup_transform(model::transform_id) const;
+
+    // Lookup transforms for input topics.
+    absl::flat_hash_map<model::transform_id, model::transform_metadata>
+      lookup_transforms_by_input_topic(model::topic_namespace_view) const;
+
+    // Get a snapshot of all the transforms that currently exist.
+    absl::flat_hash_map<model::transform_id, model::transform_metadata>
```

Comment:
btreemap ? this will cause mem pressures to have linearized memory. I'd change all the flat hash for btree maps

Reply:
Switched.